app/models/matching/RegistrationInfo.scala

```diff
@@ -16,35 +16,71 @@
 
 package models.matching
 
-import models.matching.MatchingType.{AsIndividual, AsOrganisation}
 import models.register.response.details.AddressResponse
-import play.api.libs.json.{__, OFormat, OWrites, Reads}
+import play.api.libs.json.{__, JsError, JsPath, JsSuccess, JsValue, Json, JsonValidationError, OFormat, OWrites, Reads, Writes}
 
-case class RegistrationInfo(safeId: String, name: Option[String], address: Option[AddressResponse], matchedAs: MatchingType) {
-
-  val isBusiness: Boolean   = matchedAs == AsOrganisation
-  val isIndividual: Boolean = matchedAs == AsIndividual
+sealed trait RegistrationInfo {
+  val safeId: String
 }
 
 object RegistrationInfo {
 
+  implicit val writes: Writes[RegistrationInfo] = Writes[RegistrationInfo] {
+    case ind: IndRegistrationInfo => Json.toJson(ind)(IndRegistrationInfo.writes)
+    case org: OrgRegistrationInfo => Json.toJson(org)(OrgRegistrationInfo.writes)
+  }
+
+  implicit val reads: Reads[RegistrationInfo] = Reads {
+    (data: JsValue) =>
+      val allErrors: Either[Seq[(JsPath, Seq[JsonValidationError])], RegistrationInfo] = for {
+        indErrs <- data.validate[IndRegistrationInfo].asEither.left
+        orgErrs <- data.validate[OrgRegistrationInfo].asEither.left
+      } yield indErrs.map(
+        pair => (pair._1, JsonValidationError("Failed to read as IndRegistrationInfo") +: pair._2)
+      ) ++
+        orgErrs.map(
+          pair => (pair._1, JsonValidationError("Failed to read as OrgRegistrationInfo") +: pair._2)
+        )
+
+      allErrors match {
+        case Right(lt)  => JsSuccess(lt)
+        case Left(errs) => JsError(errs)
+      }
+  }
+}
+
+case class OrgRegistrationInfo(safeId: String, name: Option[String], address: Option[AddressResponse]) extends RegistrationInfo
```

Comment:
Is this the correct way to implement readers?

Reply:
I think so, but it's actually not even needed, it was left over from something else I was trying to do and I forgot to delete it. I've removed it now.